Add module for creating secret provider class

jferris opened the pull request:
The goal of this module is to reduce the toil involved in mounting secrets.

We currently add these to the manifests for applications, but that involves copying and pasting names from the secrets we're creating.

This introduces a module that would manage the secretproviderclass in Terraform, generating the necessary YAML to copy SecretsManager secrets as environment variables.

This would mean the manifests only need to reference the secretproviderclass in pod definitions rather than defining the secret mappings in manifests.

It also provides an alternative approach to IAM permissions by allowing a single policy to be attached which allows access to all the given secrets. This should result in a smaller policy over all, as the IAM permissions are only listed once, rather than once per secret.

If we add an `environment_variables` output to our database and secret modules, those could be passed in directly to this module so that secret and environment variable names never need to be copied.

minaslater reviewed (approved):
I did not know this was an option. 🤯

jferris commented:
I converted this to a draft as I'd like to actually try this out somewhere before merging it.

jferris commented:
In practice, this module didn't prove to be usable as-is. Because we sometimes mount a secret into multiple clusters, we use separate modules for the state (IAM policies, etc) and config (secrets provider class).

In order to accommodate this, I've removed the IAM policy features in 258f558 and extracted them to a separate module: https://github.com/thoughtbot/terraform-aws-secrets/pull/7
